Automattic/wp-calypso
# Plans 2023: Sync column/label highlighting logic and other preliminary refactors/essentials

Fixes https://github.com/Automattic/wp-calypso/issues/77450
Branched off https://github.com/Automattic/wp-calypso/issues/78472

> In https://github.com/Automattic/wp-calypso/pull/77441 we shipped some changes to fix the "suggested" column highlights. However, the conditioning happing in the "useHighlightLabel" and "useHighlightAdjacencyMatrix" hooks has now diverged (e.g. the former checks whether a plan is available for purchase, and the latter doesn't). 
> 
> These two hooks should sync, otherwise, it can lead to edge cases with breakage - for example, the rounded borders are not being updated properly in some cases (see some relevant discussion [here](https://github.com/Automattic/wp-calypso/pull/77441#discussion_r1206800405)):
> 
> <img width="500" alt="Screenshot 2023-05-26 at 4 47 36 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/3ef9a9c6-eeef-4e93-bdfe-16ff99057b42">
> 
> Not the first time this scenario bites. Having two hooks that repeat some logic and are seldom used separately may not be ideal. So there's room for some interesting refactor to happen.

## Proposed Changes

- Several layers of refactors to transition things away from multiple hooks defining the highlighting logic
- Invests into a bit of refactoring to surface a cleaner API for working with the plan definitions
    - cleans up the intent system a little. a single default plans list. will be easier to migrate to store
    - introduces `GridPlan` type and values that will also consume the plan properties in follow up work
    - grid-2023 components now have access to `planRecords: Record< PlanSlug, GridPlan >` and `visiblePlans: Planslug[]`. this is _not_ the final form, although it's a better base to allow/ease further evolution
    - you'll notice a couple of `npm-ready` folders (where I'd like everything to eventually/gradually live in until package extraction).

There is a lot going on in this PR. It's another intermediary step to the larger setup we will aim for. My opinion is we focus on testing thoroughly, less on code NITs, and aim to ship ASAP. A lot of things will change in follow-up work. The current form is always better than previous though.

### Media

Some visual/UI fixes with highlight labels.

Visiting with suggested label and a lesser/free plan (`plans/yearly/[FREE]?plan=business-bundle`):

#### Before

<img width="700" alt="Screenshot 2023-07-07 at 2 54 30 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/bb7f037c-098e-436d-9c7d-ff7c091d0172">


#### After

<img width="700" alt="Screenshot 2023-07-07 at 2 56 45 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/69cb42b2-9cb6-4816-a32b-1bb44b5747a3">


## Testing Instructions



- Confirm `/start/plans` and `/plans/[paid and free]` work as before.
- Try a tailored flow e.g. `/setup/newsletter`
- Try `/setup/wooepxress` and then `plans/[woo site]` that everything works 
- Try with `?plan=business-bundle` query string and observe `suggested` highlight on column
- Confirm all the column labels and border radiuses render correctly.
- Confirm plan type selector works correctly in both Woo and Regular sites.

## Pre-merge Checklist
